File: pages/refactorings/replace-mutable-fields-with-lenses/02-player-position/Before.hs
-- Move the player inside a game: the nested copy-update touches
-- both records at every move.
module Before where

data Player = Player { pid :: String, px :: Int, py :: Int }
data Game   = Game   { level :: Int, gplayer :: Player }

moveX :: Game -> Int -> Game
moveX g dx = g { gplayer = (gplayer g) { px = px (gplayer g) + dx } }

moveY :: Game -> Int -> Game
moveY g dy = g { gplayer = (gplayer g) { py = py (gplayer g) + dy } }

File: pages/refactorings/replace-mutable-fields-with-lenses/02-player-position/Before.scala
// Move the player inside a game: the nested copy-update touches
// both records at every move.
object Before:
  case class Player(id: String, x: Int, y: Int)
  case class Game(level: Int, player: Player)

  def moveX(g: Game, dx: Int): Game =
    g.copy(player = g.player.copy(x = g.player.x + dx))

  def moveY(g: Game, dy: Int): Game =
    g.copy(player = g.player.copy(y = g.player.y + dy))
